### PR TITLE
[db]fix counting_index data race

### PR DESCRIPTION
**Pull request by millken**
# Description
When testing the analyser, it was found that db counting_index.go has a data race
```
==================
WARNING: DATA RACE
Write at 0x00c000127e60 by goroutine 36:
  github.com/iotexproject/iotex-core/db.(*countingIndex).addBatch()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/db/counting_index.go:140 +0x3fa
  github.com/iotexproject/iotex-core/db.(*countingIndex).Add()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/db/counting_index.go:118 +0x77e
  github.com/iotexproject/iotex-core/blockchain/filedao.addOneEntryToBatch()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/filedao/filedao_v2_util.go:126 +0xa4
  github.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAOv2).putBlock()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/filedao/filedao_v2_util.go:107 +0x3ed
  github.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAOv2).PutBlock()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/filedao/filedao_v2.go:236 +0x11d
  github.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAO).PutBlock()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/filedao/filedao.go:312 +0x20e
  github.com/iotexproject/iotex-core/blockchain/blockdao.(*blockDAO).PutBlock()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/blockdao/blockdao.go:244 +0x127


Previous read at 0x00c000127e60 by goroutine 62:
  github.com/iotexproject/iotex-core/db.(*countingIndex).Size()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/db/counting_index.go:112 +0x34
  github.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAOv2).getBlockStore()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/filedao/filedao_v2_util.go:183 +0x151
  github.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAOv2).GetReceipts()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/filedao/filedao_v2.go:197 +0x56
  github.com/iotexproject/iotex-core/blockchain/filedao.(*fileDAO).GetReceipts()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/filedao/filedao.go:270 +0x15c
  github.com/iotexproject/iotex-core/blockchain/blockdao.(*blockDAO).GetReceipts()
      /home/millken/.gvm/pkgsets/go1.18.6/global/pkg/mod/github.com/iotexproject/iotex-core@v1.8.3-rc4/blockchain/blockdao/blockdao.go:229 +0x13e

```

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Liuhaai** [comment]
@dustinxie

**Review by dustinxie — APPROVED** [review]

File: db/counting_index.go [hunk 1]
```diff
@@ -81,11 +82,12 @@ func NewCountingIndexNX(kv KVStore, name []byte) (CountingIndex, error) {
 		total = ZeroIndex
 	}
 
-	return &countingIndex{
+	idx := &countingIndex{
```
Comment (huangzhiran) [inline db/counting_index.go:85]: do not need atomic in initializing

File: db/counting_index.go [hunk 2]
```diff
@@ -100,16 +102,17 @@ func GetCountingIndex(kv KVStore, name []byte) (CountingIndex, error) {
...
 		return nil, errors.Wrapf(err, "counting index 0x%x doesn't exist", name)
 	}
-	return &countingIndex{
+	idx := &countingIndex{
 		kvStore: kvRange,
 		bucket:  bucket,
-		size:    byteutil.BytesToUint64BigEndian(total),
-	}, nil
+	}
+	atomic.StoreUint64(&idx.size, byteutil.BytesToUint64BigEndian(total))
```
Comment (huangzhiran) [inline db/counting_index.go:109]: do not need atomic in initializing

File: db/counting_index.go [hunk 3]
```diff
@@ -121,13 +124,14 @@ func (c *countingIndex) Add(value []byte, inBatch bool) error {
 		return errors.Wrap(ErrInvalid, "cannot call Add in batch mode, call Commit() first to exit batch mode")
 	}
 	b := batch.NewBatch()
-	b.Put(c.bucket, byteutil.Uint64ToBytesBigEndian(c.size), value, fmt.Sprintf("failed to add %d-th item", c.size+1))
-	b.Put(c.bucket, CountKey, byteutil.Uint64ToBytesBigEndian(c.size+1), fmt.Sprintf("failed to update size = %d", c.size+1))
+	size := atomic.LoadUint64(&c.size)
```
Comment (huangzhiran) [inline db/counting_index.go:127]: size := c.Size()

File: db/counting_index.go [hunk 4]
```diff
@@ -136,22 +140,23 @@ func (c *countingIndex) addBatch(value []byte) error {
...
+	c.batch.Put(c.bucket, byteutil.Uint64ToBytesBigEndian(size), value, fmt.Sprintf("failed to add %d-th item", size+1))
+	atomic.AddUint64(&c.size, 1)
 	return nil
 }
 
 // Get return value of key[slot]
 func (c *countingIndex) Get(slot uint64) ([]byte, error) {
-	if slot >= c.size {
+	if slot >= atomic.LoadUint64(&c.size) {
```
Comment (huangzhiran) [inline db/counting_index.go:151]: c.Size()

File: db/counting_index.go [hunk 5]
```diff
@@ -100,16 +102,17 @@ func GetCountingIndex(kv KVStore, name []byte) (CountingIndex, error) {
...
-	}, nil
+	}
+	atomic.StoreUint64(&idx.size, byteutil.BytesToUint64BigEndian(total))
+	return idx, nil
 }
 
 // Size returns the total number of keys so far
 func (c *countingIndex) Size() uint64 {
-	return c.size
+	return atomic.LoadUint64(&c.size)
```
Comment (huangzhiran) [inline db/counting_index.go:115]: add another func

func (c *countingIndex) Add(n uint64)  {
atomic.StoreUint64(&c.size, c.Size()+n)
}

read use c.Size(), update use c.Add(). maybe better

**Review by huangzhiran — APPROVED** [review]

**Pull request merged**